tests/e2e/cluster_downgrade_test.go

```diff
@@ -207,6 +207,88 @@ func testDowngradeUpgrade(t *testing.T, numberOfMembersToDowngrade int, clusterS
 	assert.Equal(t, beforeMembers.Members, afterMembers.Members)
 }
 
+func TestAutomaticDowngradeCancellationAfterCompletingDowngradingInClusterOf3(t *testing.T) {
+	clusterSize := 3
+
+	currentEtcdBinary := e2e.BinPath.Etcd
+	lastReleaseBinary := e2e.BinPath.EtcdLastRelease
+	if !fileutil.Exist(lastReleaseBinary) {
+		t.Skipf("%q does not exist", lastReleaseBinary)
+	}
+
+	currentVersion, err := e2e.GetVersionFromBinary(currentEtcdBinary)
+	require.NoError(t, err)
+	// wipe any pre-release suffix like -alpha.0 we see commonly in builds
+	currentVersion.PreRelease = ""
+
+	lastVersion, err := e2e.GetVersionFromBinary(lastReleaseBinary)
+	require.NoError(t, err)
+
+	require.Equalf(t, lastVersion.Minor, currentVersion.Minor-1, "unexpected minor version difference")
+	currentVersionStr := currentVersion.String()
+	lastVersionStr := lastVersion.String()
+
+	lastClusterVersion := semver.New(lastVersionStr)
+	lastClusterVersion.Patch = 0
+
+	e2e.BeforeTest(t)
+
+	t.Logf("Create cluster with version %s", currentVersionStr)
+	var snapshotCount uint64 = 10
+	epc := newCluster(t, clusterSize, snapshotCount)
+	for i := 0; i < len(epc.Procs); i++ {
+		e2e.ValidateVersion(t, epc.Cfg, epc.Procs[i], version.Versions{
+			Cluster: currentVersionStr,
+			Server:  version.Version,
+			Storage: currentVersionStr,
+		})
+	}
+	cc := epc.Etcdctl()
+	t.Logf("Cluster created")
+	if len(epc.Procs) > 1 {
+		t.Log("Waiting health interval to required to make membership changes")
+		time.Sleep(etcdserver.HealthInterval)
+	}
+
+	t.Log("Adding member to test membership, but a learner avoid breaking quorum")
+	resp, err := cc.MemberAddAsLearner(context.Background(), "fake1", []string{"http://127.0.0.1:1001"})
+	require.NoError(t, err)
+	t.Log("Removing learner to test membership")
+	_, err = cc.MemberRemove(context.Background(), resp.Member.ID)
+	require.NoError(t, err)
+	beforeMembers, beforeKV := getMembersAndKeys(t, cc)
+
+	e2e.DowngradeEnable(t, epc, lastVersion)
+
+	t.Logf("Starting downgrade process for node 0 to %q", lastVersionStr)
+	err = e2e.DowngradeUpgradeMembersByID(t, nil, epc, []int{0}, currentVersion, lastClusterVersion)
+	require.NoError(t, err)
+
+	e2e.DowngradeCancel(t, epc)
```

Comment:
This case is to test auto downgrade cancel as mentioned in https://github.com/etcd-io/etcd/pull/19365#issuecomment-2650053733, then we don't need to manually cancel it.
```suggestion
```

Reply:
I thought the idea is that
- we start the downgrade
- half way through (e.g x nodes got downgraded), we cancel the downgrade
- we continue on the downgrade
- we finish the downgrade process
- we check that indeed, there isn't anymore downgrade inflight, so the cancellation will fail with error "no inflight cancellation"

Does this make sense to you?

We can add a flag to test both the case you mentioned and the existing PR's case! What do you think? :)

Reply:
> half way through (e.g x nodes got downgraded), we cancel the downgrade

it makes this new e2e test case pointless.

Reply:
I think it's not pointless! 

My take is that if the downgrade cancellation is triggered, we can still finish the downgrade if we proceed to do so! This e2e test verifies this is actually the case.

WDYT?

Reply:
> My take is that if the downgrade cancellation is triggered, we can still finish the downgrade

It's true.

But this test case is to verify the downgrade can be automatically cancelled. If you manually cancel it, the final validation `DowngradeAutoCancelCheck` will always pass for sure. what's the point?

Reply:
> > My take is that if the downgrade cancellation is triggered, we can still finish the downgrade
> 
> It's true.

Also FYI. it works in the case of downgrade from 3.6 --> 3.5. But it's problematic if there is any protoc change; it's dangerous to continue to downgrade if the schema hasn't been migrated due to the downgrade cancellation.

Reply:
Suggest you to do some archeology (try to find out the original downgrade issue / PRs/ docs), to have a better understanding on the downgrade feature.